src/company_cmd.cpp

```diff
@@ -1170,6 +1171,88 @@ CommandCost CmdSetCompanyColour(DoCommandFlags flags, LiveryScheme scheme, bool
 	return CommandCost();
 }
 
+/**
+ * Change visibility of the rail type in build rail window.
+ * @param flags operation to perform
+ * @param type rail/road/tram type to set
+ * @param is_hidden the new visibility
+ * @return the cost of this operation or an error
+ */
+CommandCost CmdSetRailRoadTypeCompanyHidden(DoCommandFlags flags, GrfSpecFeature feature, uint8_t type, bool is_hidden)
+{
+	Company *c = Company::Get(_current_company);
+
+	switch (feature) {
+		case GSF_RAILTYPES: {
+			RailType rt = (RailType)type;
+			if (rt >= RAILTYPE_END) return CMD_ERROR;
+
+			if (flags.Test(DoCommandFlag::Execute)) {
+				c->hidden_railtypes.Set(rt, is_hidden);
+				InvalidateWindowClassesData(WC_DROPDOWN_MENU);
+			}
+
+			break;
+		}
+		case GSF_ROADTYPES:
+		case GSF_TRAMTYPES: {
+			RoadType rt = (RoadType)type;
+			if (rt >= ROADTYPE_END) return CMD_ERROR;
+
+			if (flags.Test(DoCommandFlag::Execute)) {
+				c->hidden_roadtypes.Set(rt, is_hidden);
+				InvalidateWindowClassesData(WC_DROPDOWN_MENU);
+			}
+
+			break;
+		}
+		default: NOT_REACHED();
+	}
+
+	return CommandCost();
+}
```

Comment:
I do not like using NewGRF spec feature as part of the Command "API", nor do I like passing rail type or road type as `uint8_t`.

If you use a function with the signature:

`CommandCost CmdSetRailRoadTypeCompanyHidden(DoCommandFlags flags, RailTypes railtypes, RoadTypes roadtypes, bool is_hidden)`

Then you can handle everything without needing the switch block, and the code should be shorter.